pkg/controller/dataimportcron-controller.go

```diff
@@ -0,0 +1,638 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+limitations under the License.
+See the License for the specific language governing permissions and
+*/
+
+package controller
+
+import (
+	"context"
+	"sort"
+	"time"
+
+	"github.com/go-logr/logr"
+	"github.com/gorhill/cronexpr"
+	imagev1 "github.com/openshift/api/image/v1"
+	"github.com/pkg/errors"
+
+	batchv1 "k8s.io/api/batch/v1"
+	v1beta1 "k8s.io/api/batch/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+	"kubevirt.io/containerized-data-importer/pkg/common"
+	"kubevirt.io/containerized-data-importer/pkg/util"
+)
+
+// DataImportCronReconciler members
+type DataImportCronReconciler struct {
+	client       client.Client
+	recorder     record.EventRecorder
+	scheme       *runtime.Scheme
+	log          logr.Logger
+	image        string
+	pullPolicy   string
+	cdiNamespace string
+}
+
+const (
+	// AnnSourceDesiredDigest is the digest of the pending updated image
+	AnnSourceDesiredDigest = AnnAPIGroup + "/storage.import.sourceDesiredDigest"
+	// AnnSourceUpdatePending indicates the source image digest was updated, and the image is pending for import based on the cron schedule
+	AnnSourceUpdatePending = AnnAPIGroup + "/storage.import.sourceUpdatePending"
+	// AnnCronInitialized indicates the cron was initialized
+	AnnCronInitialized = AnnAPIGroup + "/storage.import.cronInitialized"
+	// AnnNextCronTime is the next time stamp which satisfies the cron expression
+	AnnNextCronTime = AnnAPIGroup + "/storage.import.nextCronTime"
+
+	// dataImportCronFinalizer ensures CronJob is deleted when DataImportCron is deleted, as there is no cross-namespace OwnerReference
+	dataImportCronFinalizer = "cdi.kubevirt.io/dataImportCronFinalizer"
+
+	dataImportControllerName      = "dataimportcron-controller"
+	labelDataImportCronName       = "dataimportcron-name"
+	recentImportsToKeepPerCronJob = 3
+	digestPrefix                  = "sha256:"
+	digestDvNameSuffixLength      = 12
+)
+
+// Reconcile loop for DataImportCronReconciler
+func (r *DataImportCronReconciler) Reconcile(ctx context.Context, req reconcile.Request) (reconcile.Result, error) {
```

Comment:
This is bad form.  A controller should only reconcile a single resource type.  What if there is a DIC and a ImageStream in the same namespace with the same name?

The ImageStream watch should look up all DICs that refer to theImageStream and queue those

Reply:
Nice point, which could be solved much simper if it was 1:1 DIC:IS. So with your suggestion on every DIC reconcile, if it has an IS source, it must Get() it and check the digest.

Reply:
Yeah, if the DIC is requeued to be in accordance with cron schedule I'm not sure you even have to watch ImageStreams

Reply:
Just reminding that not-watching-IS and just getting it when its cron is scheduled was my initial suggestion, and someone opposed...

Reply:
Pushed both options (top patch removes ImageStream watch).